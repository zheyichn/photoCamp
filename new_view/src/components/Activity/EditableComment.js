import React, { useState } from "react";
import { Avatar, Button, Grid, Typography } from "@mui/material";
import CustomInput from "./CustomInput";
import { deleteComment, updateComment } from "../../api/mock_api_k";
import DeleteOutlineIcon from "@mui/icons-material/DeleteOutline";
import ModeEditOutlineOutlinedIcon from "@mui/icons-material/ModeEditOutlineOutlined";
import {
  editCommentButtonStyle,
  editCommentIconStyle,
} from "../../styles/buttonStyles";

export default function EditableComment(props) {
  const {
    commentText,
    commentorName,
    commentorProfile,
    createdAt,
    postId,
    commentId,
    commentorId,
    reFetchComments,
  } = props;
  const [value, setValue] = useState(commentText);
  const [edit, setEdit] = useState(false);
  const formattedTime = new Date(createdAt).toLocaleDateString("en-US");

  const handleEdit = () => {
    setEdit(true);
  };

  const handleDelete = async () => {
    try {
      await deleteComment(postId, commentId);
      reFetchComments(postId);
    } catch (e) {
      console.error(e);
    }
  };

  const handleSave = async (val) => {
    setValue(val); // now the value state stores the comment after editing
    const newCommentObj = {
      commentText: val,
      commentorId: commentorId,
      createdAt: createdAt,
    };
    try {
      await updateComment(commentId, newCommentObj);
      setEdit(false);
    } catch (e) {
      console.error(e);
    }
  };

  const close = () => {
    setEdit(false);
  };

  return (
    <Grid
      container
      flexDirection="row"
      justifyContent="flex-start"
      alignItems="center"
      mb={2}
    >
      <Grid container item justifyContent="flex-start">
        <Grid item>
          <Avatar src={commentorProfile} alt={commentorName} />
        </Grid>
        <Grid item pl={1} pt={1}>
          <Typography data-testid="test-name" variant="subtitle1">
            {commentorName ? commentorName : "userName"}
          </Typography>
        </Grid>
        <Grid item pl={1} pt={1.5}>
          <Typography sx={{ fontSize: "14px" }}>{formattedTime}</Typography>
        </Grid>
      </Grid>

      <Grid item container ml={1} mt={1} flexDirection="column">
        {!edit ? (
          <>
            <Grid item>
              <Typography className="myComment">{value}</Typography>
            </Grid>
            <Grid alignSelf="flex-end">
              <Button
                data-testid="edit-btn"
                onClick={handleEdit}
                style={editCommentButtonStyle}
              >
                <ModeEditOutlineOutlinedIcon style={editCommentIconStyle} />
              </Button>
              <Button
                className="deleteCyComment"
                data-testid="delete-btn"
                onClick={handleDelete}
                style={editCommentButtonStyle}
              >
                <DeleteOutlineIcon style={editCommentIconStyle} />
              </Button>
            </Grid>
          </>
        ) : (
          <CustomInput
            data-testid="input-test"
            defaultValue={value}
            saveText={handleSave}
            cancelEdit={close}
          />
        )}
      </Grid>
    </Grid>
  );
}
