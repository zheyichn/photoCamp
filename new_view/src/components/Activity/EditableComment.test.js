import React from "react";
import { render, screen, cleanup } from "@testing-library/react";
import EditableComment from "./EditableComment";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";

afterEach(() => {
  cleanup();
});

describe("EditableComment component renders correctly", () => {
  test("buttons displayed correctly", () => {
    render(<EditableComment />);
    const editButton = screen.getByTestId("edit-btn");
    const deleteButton = screen.getByTestId("delete-btn");
    expect(editButton).toBeInTheDocument();
    expect(deleteButton).toBeInTheDocument();
  });

  test("commentorName displays appropriately", () => {
    const fakeCommentorName = "Daniel";
    render(<EditableComment commentorName={fakeCommentorName} />);
    const userName = screen.getByTestId("test-name");
    expect(userName).toHaveTextContent("Daniel");
  });

  test("cinput section not shown by default", () => {
    const { queryByTestId } = render(<EditableComment />);
    expect(queryByTestId("input-test")).toBeNull();
  });

  it("editableComment match snapshots", () => {
    const tree = renderer.create(<EditableComment />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
