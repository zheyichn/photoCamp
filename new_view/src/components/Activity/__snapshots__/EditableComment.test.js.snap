// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableComment component renders correctly editableComment match snapshots 1`] = `
<div
  className="MuiGrid-root MuiGrid-container css-yto8eg-MuiGrid-root"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-item css-19hkhe-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <div
        className="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-2s90m6-MuiAvatar-root"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
          data-testid="PersonIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
          />
        </svg>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item css-1629zra-MuiGrid-root"
    >
      <h6
        className="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
        data-testid="test-name"
      >
        userName
      </h6>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item css-17apry7-MuiGrid-root"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 css-1d9ye53-MuiTypography-root"
      >
        Invalid Date
      </p>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-item css-fzvh25-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      />
    </div>
    <div
      className="MuiGrid-root css-1lh7sv1-MuiGrid-root"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        data-testid="edit-btn"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          {
            "maxHeight": "30px",
            "maxWidth": "15x",
            "minHeight": "30px",
            "minWidth": "15px",
          }
        }
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ModeEditOutlineOutlinedIcon"
          focusable="false"
          style={
            {
              "color": "#3AB4F2",
              "maxHeight": "18px",
              "maxWidth": "18x",
              "minHeight": "18px",
              "minWidth": "18px",
            }
          }
          viewBox="0 0 24 24"
        >
          <path
            d="M3 21h3.75L17.81 9.94l-3.75-3.75L3 17.25V21zm2-2.92 9.06-9.06.92.92L5.92 19H5v-.92zM18.37 3.29a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83c.39-.39.39-1.02 0-1.41l-2.34-2.34z"
          />
        </svg>
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        data-testid="delete-btn"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          {
            "maxHeight": "30px",
            "maxWidth": "15x",
            "minHeight": "30px",
            "minWidth": "15px",
          }
        }
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="DeleteOutlineIcon"
          focusable="false"
          style={
            {
              "color": "#3AB4F2",
              "maxHeight": "18px",
              "maxWidth": "18x",
              "minHeight": "18px",
              "minWidth": "18px",
            }
          }
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM8 9h8v10H8V9zm7.5-5-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
